File: orchestrator/tests/test-db.js
import { describe, it, before, after } from 'node:test';
import assert from 'node:assert/strict';
import { mkdtempSync, rmSync } from 'fs';
import { join } from 'path';
import { tmpdir } from 'os';
import {
  initDb, getDb, closeDb,
  queueEnqueue, queueDequeue, queuePeek, queueRemove, queueList,
  projectCreate, projectUpdate, projectGet, projectGetById, projectsByStatus,
  portAllocate, portDeallocate, portGetForProject,
  configGet, configSet, configGetBool,
  ideasForDate, ideaGet
} from '../src/db.js';

let tmpDir;

before(() => {
  tmpDir = mkdtempSync(join(tmpdir(), 'nsaf-test-'));
  initDb(join(tmpDir, 'test.db'));
});

after(() => {
  closeDb();
  rmSync(tmpDir, { recursive: true, force: true });
});

describe('Schema initialization', () => {
  it('creates all tables', () => {
    const db = getDb();
    const tables = db.prepare("SELECT name FROM sqlite_master WHERE type='table' ORDER BY name").all();
    const names = tables.map(t => t.name);
    assert.ok(names.includes('ideas'));
    assert.ok(names.includes('idea_history'));
    assert.ok(names.includes('projects'));
    assert.ok(names.includes('queue'));
    assert.ok(names.includes('ports'));
    assert.ok(names.includes('config'));
  });

  it('uses WAL mode', () => {
    const db = getDb();
    const mode = db.pragma('journal_mode', { simple: true });
    assert.equal(mode, 'wal');
  });
});

describe('Config operations', () => {
  it('get returns null for missing key', () => {
    assert.equal(configGet('nonexistent'), null);
  });

  it('set and get', () => {
    configSet('test_key', 'test_value');
    assert.equal(configGet('test_key'), 'test_value');
  });

  it('set overwrites existing', () => {
    configSet('test_key', 'new_value');
    assert.equal(configGet('test_key'), 'new_value');
  });

  it('getBool returns correct boolean', () => {
    configSet('flag_true', 'true');
    configSet('flag_false', 'false');
    configSet('flag_one', '1');
    assert.equal(configGetBool('flag_true'), true);
    assert.equal(configGetBool('flag_false'), false);
    assert.equal(configGetBool('flag_one'), true);
    assert.equal(configGetBool('missing_flag'), false);
  });
});

describe('Project operations', () => {
  let projectId;

  it('creates a project', () => {
    projectId = projectCreate({ slug: 'test-app', ideaId: null, projectDir: '/tmp/test-app' });
    assert.ok(projectId > 0);
  });

  it('gets project by slug', () => {
    const p = projectGet('test-app');
    assert.equal(p.slug, 'test-app');
    assert.equal(p.status, 'queued');
    assert.equal(p.project_dir, '/tmp/test-app');
  });

  it('gets project by id', () => {
    const p = projectGetById(projectId);
    assert.equal(p.slug, 'test-app');
  });

  it('returns null for missing project', () => {
    assert.equal(projectGet('nonexistent'), null);
  });

  it('updates project fields', () => {
    projectUpdate('test-app', { status: 'building', sdd_phase: 'design' });
    const p = projectGet('test-app');
    assert.equal(p.status, 'building');
    assert.equal(p.sdd_phase, 'design');
  });

  it('lists projects by status', () => {
    const building = projectsByStatus('building');
    assert.ok(building.length >= 1);
    assert.ok(building.every(p => p.status === 'building'));
  });
});

describe('Queue operations', () => {
  let pid1, pid2, pid3;

  before(() => {
    pid1 = projectCreate({ slug: 'q-app-1', ideaId: null, projectDir: '/tmp/q1' });
    pid2 = projectCreate({ slug: 'q-app-2', ideaId: null, projectDir: '/tmp/q2' });
    pid3 = projectCreate({ slug: 'q-app-3', ideaId: null, projectDir: '/tmp/q3' });
  });

  it('enqueues projects in order', () => {
    queueEnqueue(pid1);
    queueEnqueue(pid2);
    queueEnqueue(pid3);
    const list = queueList();
    assert.equal(list.length, 3);
    assert.equal(list[0].id, pid1);
    assert.equal(list[1].id, pid2);
    assert.equal(list[2].id, pid3);
  });

  it('peek returns first without removing', () => {
    const peeked = queuePeek();
    assert.equal(peeked.project_id, pid1);
    assert.equal(queueList().length, 3);
  });

  it('dequeue returns and removes first', () => {
    const item = queueDequeue();
    assert.equal(item.project_id, pid1);
    assert.equal(queueList().length, 2);
  });

  it('remove by project id', () => {
    queueRemove(pid2);
    const list = queueList();
    assert.equal(list.length, 1);
    assert.equal(list[0].id, pid3);
  });

  it('dequeue returns null when empty', () => {
    queueDequeue(); // remove pid3
    assert.equal(queueDequeue(), null);
  });
});

describe('Port operations', () => {
  let portProjectId;

  before(() => {
    portProjectId = projectCreate({ slug: 'port-app', ideaId: null, projectDir: '/tmp/port' });
  });

  it('allocates a port batch', () => {
    const ports = portAllocate(portProjectId, 5020, 5999, 10);
    assert.equal(ports.portStart, 5020);
    assert.equal(ports.portEnd, 5029);
  });

  it('get ports for project', () => {
    const ports = portGetForProject(portProjectId);
    assert.equal(ports.port_start, 5020);
    assert.equal(ports.port_end, 5029);
  });

  it('allocates next available batch', () => {
    const pid2 = projectCreate({ slug: 'port-app-2', ideaId: null, projectDir: '/tmp/port2' });
    const ports = portAllocate(pid2, 5020, 5999, 10);
    assert.equal(ports.portStart, 5030);
    assert.equal(ports.portEnd, 5039);
  });

  it('deallocates and reclaims', () => {
    portDeallocate(portProjectId);
    assert.equal(portGetForProject(portProjectId), null);
    const pid3 = projectCreate({ slug: 'port-app-3', ideaId: null, projectDir: '/tmp/port3' });
    const ports = portAllocate(pid3, 5020, 5999, 10);
    assert.equal(ports.portStart, 5020);
  });

  it('throws when range exhausted', () => {
    const pid = projectCreate({ slug: 'port-exhaust', ideaId: null, projectDir: '/tmp/pe' });
    assert.throws(() => portAllocate(pid, 5020, 5029, 10), /No available port batches/);
  });
});

describe('Idea operations', () => {
  it('reads ideas for date', () => {
    const db = getDb();
    db.prepare(`
      INSERT INTO ideas (date, source, rank, name, description, category, complexity, suggested_stack)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?)
    `).run('2026-04-01', 'openai', 1, 'Test App', 'A test app', 'productivity', 'medium', '{}');

    const ideas = ideasForDate('2026-04-01');
    assert.ok(ideas.length >= 1);
    assert.equal(ideas[0].name, 'Test App');
  });

  it('gets idea by id', () => {
    const ideas = ideasForDate('2026-04-01');
    const idea = ideaGet(ideas[0].id);
    assert.equal(idea.name, 'Test App');
  });

  it('returns null for missing idea', () => {
    assert.equal(ideaGet(99999), null);
  });
});

File: shared/tests/test_db.py
import os
import tempfile
import pytest

from shared.db import (
    get_db, close_db, reset_db,
    ideas_insert, ideas_insert_batch, ideas_for_date, idea_get,
    history_insert, history_insert_batch, history_all,
    project_create, project_update, project_get, projects_by_status,
    queue_enqueue, queue_remove, queue_list,
    config_get, config_set,
)


@pytest.fixture(autouse=True)
def fresh_db(tmp_path):
    db_path = str(tmp_path / "test.db")
    db = reset_db(db_path)
    # Create schema
    db.executescript("""
        CREATE TABLE IF NOT EXISTS ideas (
            id INTEGER PRIMARY KEY,
            date TEXT NOT NULL,
            source TEXT NOT NULL,
            rank INTEGER NOT NULL,
            name TEXT NOT NULL,
            description TEXT NOT NULL,
            category TEXT NOT NULL,
            complexity TEXT NOT NULL,
            suggested_stack TEXT,
            selected INTEGER DEFAULT 0,
            created_at TEXT DEFAULT (datetime('now'))
        );
        CREATE TABLE IF NOT EXISTS idea_history (
            id INTEGER PRIMARY KEY,
            name TEXT NOT NULL,
            description TEXT NOT NULL,
            date TEXT NOT NULL
        );
        CREATE TABLE IF NOT EXISTS projects (
            id INTEGER PRIMARY KEY,
            idea_id INTEGER REFERENCES ideas(id),
            slug TEXT UNIQUE NOT NULL,
            status TEXT NOT NULL DEFAULT 'queued',
            port_start INTEGER,
            port_end INTEGER,
            db_name TEXT,
            project_dir TEXT NOT NULL,
            sdd_phase TEXT,
            sdd_active_role TEXT,
            sdd_progress INTEGER DEFAULT 0,
            deployed_url TEXT,
            render_url TEXT,
            last_state_change TEXT,
            stall_alerted INTEGER DEFAULT 0,
            started_at TEXT,
            completed_at TEXT,
            created_at TEXT DEFAULT (datetime('now'))
        );
        CREATE TABLE IF NOT EXISTS queue (
            id INTEGER PRIMARY KEY,
            project_id INTEGER REFERENCES projects(id),
            position INTEGER NOT NULL,
            created_at TEXT DEFAULT (datetime('now'))
        );
        CREATE TABLE IF NOT EXISTS ports (
            port_start INTEGER PRIMARY KEY,
            port_end INTEGER NOT NULL,
            project_id INTEGER REFERENCES projects(id),
            allocated_at TEXT DEFAULT (datetime('now'))
        );
        CREATE TABLE IF NOT EXISTS config (
            key TEXT PRIMARY KEY,
            value TEXT NOT NULL
        );
        CREATE INDEX IF NOT EXISTS idx_ideas_date ON ideas(date);
        CREATE INDEX IF NOT EXISTS idx_projects_status ON projects(status);
        CREATE INDEX IF NOT EXISTS idx_queue_position ON queue(position);
    """)
    yield db
    close_db()


class TestIdeas:
    def test_insert_and_get(self):
        idea = {
            "date": "2026-04-01",
            "source": "openai",
            "rank": 1,
            "name": "Fitness Tracker",
            "description": "Track workouts",
            "category": "fitness",
            "complexity": "medium",
            "suggested_stack": '{"frontend": "React"}',
        }
        idea_id = ideas_insert(idea)
        assert idea_id > 0

        fetched = idea_get(idea_id)
        assert fetched["name"] == "Fitness Tracker"
        assert fetched["source"] == "openai"

    def test_insert_batch(self):
        ideas = [
            {
                "date": "2026-04-01", "source": "gemini", "rank": i,
                "name": f"App {i}", "description": f"Desc {i}",
                "category": "productivity", "complexity": "low",
                "suggested_stack": None,
            }
            for i in range(1, 11)
        ]
        ideas_insert_batch(ideas)
        fetched = ideas_for_date("2026-04-01")
        assert len(fetched) == 10

    def test_ideas_for_date_empty(self):
        assert ideas_for_date("2099-01-01") == []

    def test_idea_get_missing(self):
        assert idea_get(99999) is None


class TestIdeaHistory:
    def test_insert_and_list(self):
        history_insert("Old App", "An old app", "2026-03-01")
        items = history_all()
        assert len(items) == 1
        assert items[0]["name"] == "Old App"

    def test_insert_batch(self):
        batch = [
            {"name": f"Past {i}", "description": f"Desc {i}", "date": "2026-03-01"}
            for i in range(5)
        ]
        history_insert_batch(batch)
        assert len(history_all()) == 5


class TestProjects:
    def test_create_and_get(self):
        pid = project_create("my-app", None, "/tmp/my-app")
        assert pid > 0
        p = project_get("my-app")
        assert p["slug"] == "my-app"
        assert p["status"] == "queued"

    def test_get_missing(self):
        assert project_get("nonexistent") is None

    def test_update(self):
        project_create("upd-app", None, "/tmp/upd")
        project_update("upd-app", status="building", sdd_phase="design")
        p = project_get("upd-app")
        assert p["status"] == "building"
        assert p["sdd_phase"] == "design"

    def test_by_status(self):
        project_create("s1", None, "/tmp/s1")
        project_create("s2", None, "/tmp/s2")
        project_update("s1", status="building")
        building = projects_by_status("building")
        assert any(p["slug"] == "s1" for p in building)
        queued = projects_by_status("queued")
        assert any(p["slug"] == "s2" for p in queued)


class TestQueue:
    def test_enqueue_and_list(self):
        pid1 = project_create("qa1", None, "/tmp/qa1")
        pid2 = project_create("qa2", None, "/tmp/qa2")
        queue_enqueue(pid1)
        queue_enqueue(pid2)
        q = queue_list()
        assert len(q) == 2
        assert q[0]["slug"] == "qa1"
        assert q[1]["slug"] == "qa2"

    def test_remove(self):
        pid = project_create("qr1", None, "/tmp/qr1")
        queue_enqueue(pid)
        queue_remove(pid)
        assert queue_list() == []


class TestConfig:
    def test_get_missing(self):
        assert config_get("missing") is None

    def test_set_and_get(self):
        config_set("key1", "val1")
        assert config_get("key1") == "val1"

    def test_overwrite(self):
        config_set("key2", "a")
        config_set("key2", "b")
        assert config_get("key2") == "b"
